Realm/RLMSyncSubscription.mm

```diff
@@ -0,0 +1,483 @@
+                                  subscriptionName:self.name
+                                         predicate:predicate
+                                    updateExisting:true];
+}
+
+@end
+
+#pragma mark - SubscriptionSet
+
+@interface RLMSyncSubscriptionSet () {
+    std::unique_ptr<realm::sync::SubscriptionSet> _subscriptionSet;
+    std::unique_ptr<realm::sync::MutableSubscriptionSet> _mutableSubscriptionSet;
+}
+@end
+
+@implementation RLMSyncSubscriptionSet {
+    BOOL isInWriteTransaction;
+    RLMRealm *_realm;
+    id _strongBuffer[16];
+}
+
+- (instancetype)initWithSubscriptionSet:(realm::sync::SubscriptionSet)subscriptionSet
+                                  realm:(RLMRealm *)realm {
+    if (self = [super init]) {
+        _subscriptionSet = std::make_unique<realm::sync::SubscriptionSet>(subscriptionSet);
+        _realm = realm;
+        self->isInWriteTransaction = false;
+        return self;
+    }
+    return nil;
+}
+
+- (NSUInteger)count {
+    return _subscriptionSet->size();
+}
+
+- (nullable NSError *)error {
+    NSString *errorMessage = RLMStringDataToNSString(_subscriptionSet->error_str());
+    if ([errorMessage length] == 0) {
+        return NULL;
+    }
+    return [[NSError alloc]initWithDomain:RLMSyncErrorDomain
+                                               code:0
+                                           userInfo:@{NSLocalizedDescriptionKey : errorMessage}];
+}
+
+- (RLMSyncSubscriptionState)state {
+    switch (_subscriptionSet->state()) {
+        case realm::sync::SubscriptionSet::State::Uncommitted:
+        case realm::sync::SubscriptionSet::State::Pending:
+        case realm::sync::SubscriptionSet::State::Bootstrapping:
+            return RLMSyncSubscriptionStatePending;
+        case realm::sync::SubscriptionSet::State::Complete:
+            return RLMSyncSubscriptionStateComplete;
+        case realm::sync::SubscriptionSet::State::Error:
+            return RLMSyncSubscriptionStateError;
+        case realm::sync::SubscriptionSet::State::Superceded:
+            return RLMSyncSubscriptionStateSuperceded;
+    }
+}
+
+#pragma mark - Batch Update subscriptions
+
+- (void)write:(__attribute__((noescape)) void(^)(void))block {
+    return [self write:block onComplete:^(NSError**){}];
+}
+
+- (void)write:(__attribute__((noescape)) void(^)(void))block
+   onComplete:(void(^)(NSError**))completionBlock {
+    [self secureWrite];
+    _mutableSubscriptionSet = std::make_unique<realm::sync::MutableSubscriptionSet>(_subscriptionSet->make_mutable_copy());
+    self->isInWriteTransaction = true;
+    block();
+    try {
+        _subscriptionSet = std::make_unique<realm::sync::SubscriptionSet>(std::move(*_mutableSubscriptionSet).commit());
+        _mutableSubscriptionSet = nullptr;
+        self->isInWriteTransaction = false;
+    }
+    catch (std::exception error) {
+        NSError *err = [[NSError alloc] initWithDomain:@"subscription_set" code:-1 userInfo:@{@"reason":@(error.what())}];
```

Comment:
`@"subscription_set"` is not a sensible error domain to be using. The error domain should always be a constant exposed as part of the API, and the string chosen needs to have a RLM prefix.

Reply:
Fixed